blockchain/blockchain.go

```diff
@@ -0,0 +1,182 @@
+		} else {
+			agent = a
+		}
+
+		// Setup identity
+		if privateKeyString := config.GetString(config.PrivateKeyKey); privateKeyString != "" {
+			if privKey, err := crypto.HexToECDSA(privateKeyString); err != nil {
+				log.WithError(err).Panic("error getting private key")
+			} else {
+				privateKey = privKey
+				address = crypto.PubkeyToAddress(privateKey.PublicKey).Hex()
+			}
+		} else if hdwalletMnemonic := config.GetString(config.HdwalletMnemonicKey); hdwalletMnemonic != "" {
+			if privKey, err := derivePrivateKey(hdwalletMnemonic, 44, 60, 0, 0, uint32(config.GetInt(config.HdwalletIndexKey))); err != nil {
+				log.WithError(err).Panic("error deriving private key")
+			} else {
+				privateKey = privKey
+				address = crypto.PubkeyToAddress(privateKey.PublicKey).Hex()
+			}
+		}
+
+		// Start event and job completion routines
+		jobCompletionQueue = make(chan *jobInfo, 1000)
+		go processJobCompletions()
+		go processEvents()
+		go submitOldJobsForCompletion()
+	}
+}
+
+func GetGrpcStreamInterceptor() grpc.StreamServerInterceptor {
+	if config.GetBool(config.BlockchainEnabledKey) {
+		return jobValidationInterceptor
+	} else {
+		return noOpInterceptor
+	}
+}
+
+func IsValidJobInvocation(jobAddressBytes, jobSignatureBytes []byte) bool {
+	log := log.WithFields(log.Fields{
+		"jobAddress":   common.BytesToAddress(jobAddressBytes).Hex(),
+		"jobSignature": hex.EncodeToString(jobSignatureBytes)})
+
+	// Get job state from db
+	log.Debug("retrieving job from database")
+	job := &db.Job{}
+
+	db.View(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(db.JobBucketName)
+		jobBytes := bucket.Get(jobAddressBytes)
+		if jobBytes != nil {
+			json.Unmarshal(jobBytes, job)
+		}
+		return nil
+	})
+
+	// If job is marked completed locally, reject
+	if job.Completed {
+		log.Debug("job already marked completed locally")
+		return false
+	}
+
+	v, r, s, err := parseSignature(jobSignatureBytes)
+	if err != nil {
+		log.WithError(err).Error("error parsing signature")
+		return false
+	}
+
+	pubKey, err := crypto.SigToPub(crypto.Keccak256([]byte{0x19}, []byte("Ethereum Signed Message:\n32"),
+		crypto.Keccak256(jobAddressBytes)), bytes.Join([][]byte{jobSignatureBytes[0:64], {v % 27}}, []byte{}))
+	if err != nil {
+		log.WithError(err).Error("error recovering signature")
+		return false
+	}
+
+	// If job is FUNDED and signature validates, skip on-chain validation
+	if job.JobState == JobFundedState && bytes.Equal(crypto.PubkeyToAddress(*pubKey).Bytes(), job.Consumer) {
+		log.Debug("validated job invocation locally")
+		return true
+	}
+
+	log.Debug("unable to validate job invocation locally; falling back to on-chain validation")
+
+	// Fall back to on-chain validation
+	if validated, err := agent.ValidateJobInvocation(&bind.CallOpts{
+		Pending: true,
+		From:    common.HexToAddress(address)}, common.BytesToAddress(jobAddressBytes), v, r, s); err != nil {
+		log.WithError(err).Error("error validating job on chain")
+		return false
```

Comment:
this "log error and return false" makes me think that we should be returning an `error` or `nil` so that the caller can tell what happened (and log it itself; a validity check shouldn't really be logging this much)